File: python/v1/buyers/creatives/create_video_creatives.py
#!/usr/bin/python
#
# Copyright 2020 Google Inc. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Creates a creative with video content for the given buyer account ID."""


import argparse
import datetime
import os
import pprint
import sys
import uuid

sys.path.insert(0, os.path.abspath('../../..'))

from googleapiclient.errors import HttpError
import util


_BUYER_NAME_TEMPLATE = 'buyers/%s'

DEFAULT_BUYER_RESOURCE_ID = 'ENTER_BUYER_RESOURCE_ID_HERE'


def main(realtimebidding, args):
  account_id = args.account_id

  creative = {
    'advertiserName': args.advertiser_name,
    'creativeId': args.creative_id,
    'declaredAttributes': args.declared_attributes,
    'declaredClickThroughUrls': args.declared_click_urls,
    'declaredRestrictedCategories': args.declared_restricted_categories,
    'declaredVendorIds': [int(id) for id in args.declared_vendor_ids],
    'video': {
        'videoUrl': args.video_url
    }
  }

  print(f'Creating video creative for buyer account ID "{account_id}":')
  try:
    # Construct and execute the request.
    response = (realtimebidding.buyers().creatives().create(
        parent=_BUYER_NAME_TEMPLATE % account_id, body=creative).execute())
  except HttpError as e:
    print(e)
    sys.exit(1)

  pprint.pprint(response)


if __name__ == '__main__':
  try:
    service = util.GetService(version='v1')
  except IOError as ex:
    print(f'Unable to create realtimebidding service - {ex}')
    print('Did you specify the key file in util.py?')
    sys.exit(1)

  parser = argparse.ArgumentParser(
      description=('Creates a video creative for the given buyer account ID.')
  )

  # Required fields.
  parser.add_argument(
      '-a', '--account_id', default=DEFAULT_BUYER_RESOURCE_ID,
      help=('The resource ID of the buyers resource under which the '
            'creative is to be created.'))
  # Optional fields.
  parser.add_argument(
      '--advertiser_name', default='Test',
      help='The name of the company being advertised in the creative.')
  parser.add_argument(
      '-c', '--creative_id', default='HTML_Creative_%s' % uuid.uuid4(),
      help=('The user-specified creative ID. The maximum length of the '
            'creative ID is 128 bytes.'))
  parser.add_argument(
      '--declared_attributes', nargs='*', default=['CREATIVE_TYPE_VAST_VIDEO'],
      help=('The creative attributes being declared. Specify each attribute '
            'separated by a space.'))
  parser.add_argument(
      '--declared_click_urls', nargs='*', default=['http://test.com'],
      help=('The click-through URLs being declared. Specify each URL '
            'separated by a space.'))
  parser.add_argument(
      '--declared_restricted_categories', nargs='*', default=[],
      help=('The restricted categories being declared. Specify each category '
            'separated by a space.'))
  parser.add_argument(
      '--declared_vendor_ids', nargs='*', default=[],
      help=('The vendor IDs being declared. Specify each ID separated by a '
            'space.'))
  parser.add_argument(
      '--video_url', help='The URL to fetch a video ad.',
      default='https://video.test.com/ads?id=123456&wprice=%%WINNING_PRICE%%')

  args = parser.parse_args()

  main(service, args)

File: CSharp/v1/Buyers/Creatives/CreateVideoCreatives.cs
/* Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using Google.Apis.RealTimeBidding.v1;
using Google.Apis.RealTimeBidding.v1.Data;
using Mono.Options;

using System;
using System.Collections.Generic;

namespace Google.Apis.RealTimeBidding.Examples.v1.Buyers.Creatives
{
    /// <summary>
    /// Creates a creative with video content for the given buyer account ID.
    /// </summary>
    public class CreateVideoCreatives : ExampleBase
    {
        private RealTimeBiddingService rtbService;

        /// <summary>
        /// Constructor.
        /// </summary>
        public CreateVideoCreatives()
        {
            rtbService = Utilities.GetRealTimeBiddingService();
        }

        /// <summary>
        /// Returns a description about the code example.
        /// </summary>
        public override string Description
        {
            get => "This code example creates a creative with video content for the given buyer " +
                   "account ID.";
        }

        /// <summary>
        /// Parse specified arguments.
        /// </summary>
        protected override Dictionary<string, object> ParseArguments(List<string> exampleArgs) {
            string[] requiredOptions = new string[] {"account_id"};
            bool showHelp = false;

            string accountId = null;
            string advertiserName = null;
            string creativeId = null;
            IList<string> declaredAttributes = new List<string>();
            IList<string> declaredClickUrls = new List<string>();
            IList<string> declaredRestrictedCategories = new List<string>();
            IList<int?> declaredVendorIds = new List<int?>();
            string videoUrl = null;

            var defaultVideoUrl = "https://video.test.com/ads?id=123456&wprice=%%WINNING_PRICE%%";

            OptionSet options = new OptionSet {
                "Creates a creative with video content for the given buyer account ID.",
                {
                    "h|help",
                    "Show help message and exit.",
                    h => showHelp = h != null
                },
                {
                    "a|account_id=",
                    ("[Required] The resource ID of the buyers resource under which the " +
                     "creatives were created. This will be used to construct the name used as " +
                     "a path parameter for the creatives.get request."),
                    a => accountId = a
                },
                {
                    "advertiser_name=",
                    "The name of the company being advertised in the creative.",
                    advertiser_name => advertiserName = advertiser_name
                },
                {
                    "c|creative_id=",
                    ("The user-specified creative ID. The maximum length of the creative ID is " +
                     "128 bytes."),
                    c => creativeId = c
                },
                {
                    "declared_attributes=",
                    ("The creative attributes being declared. Specify this argument for each " +
                     "value you intend to include."),
                    declared_attributes => declaredAttributes.Add(declared_attributes)
                },
                {
                    "declared_click_urls=",
                    ("The click-through URLs being declared. Specify this argument for each " +
                     "value you intend to include."),
                    declared_click_through_urls => declaredClickUrls.Add(
                        declared_click_through_urls)
                },
                {
                    "declared_restricted_categories=",
                    ("The restricted categories being declared. Specify this argument for each " +
                     "value you intend to include."),
                    declared_restricted_categories => declaredRestrictedCategories.Add(
                        declared_restricted_categories)
                },
                {
                    "declared_vendor_ids=",
                    ("The vendor IDs being declared. Specify this argument for each value you " +
                     "intend to include."),
                    (int declared_vendor_ids) => declaredVendorIds.Add(declared_vendor_ids)
                },
                {
                    "video_url=",
                    "The URL used to fetch a video ad.",
                    video_url => videoUrl = video_url
                }
            };

            List<string> extras = options.Parse(exampleArgs);
            var parsedArgs = new Dictionary<string, object>();

            // Show help message.
            if(showHelp == true)
            {
                options.WriteOptionDescriptions(Console.Out);
                Environment.Exit(0);
            }
            // Set arguments.
            parsedArgs["account_id"] = accountId;
            parsedArgs["advertiser_name"] = advertiserName ?? "Test";
            parsedArgs["creative_id"] = creativeId ?? String.Format(
                "Video_Creative_{0}", System.Guid.NewGuid());

            if (declaredAttributes.Count == 0)
            {
                declaredAttributes.Add("CREATIVE_TYPE_VAST_VIDEO");
            }
            parsedArgs["declared_attributes"] = declaredAttributes;

            if (declaredClickUrls.Count == 0)
            {
                declaredClickUrls.Add("http://test.com");
            }

            parsedArgs["declared_click_urls"] = declaredClickUrls;
            parsedArgs["declared_restricted_categories"] = declaredRestrictedCategories;
            parsedArgs["declared_vendor_ids"] = declaredVendorIds;
            parsedArgs["video_url"] = videoUrl ?? defaultVideoUrl;
            // Validate that options were set correctly.
            Utilities.ValidateOptions(options, parsedArgs, requiredOptions, extras);

            return parsedArgs;
        }

        /// <summary>
        /// Run the example.
        /// </summary>
        /// <param name="parsedArgs">Parsed arguments for the example.</param>
        protected override void Run(Dictionary<string, object> parsedArgs)
        {
            string accountId = (string) parsedArgs["account_id"];
            string parent = $"buyers/{accountId}";

            VideoContent videoContent = new VideoContent();
            videoContent.VideoUrl = (string) parsedArgs["video_url"];

            Creative newCreative = new Creative();
            newCreative.AdvertiserName = (string) parsedArgs["advertiser_name"];
            newCreative.CreativeId = (string) parsedArgs["creative_id"];
            newCreative.DeclaredAttributes = (IList<string>) parsedArgs["declared_attributes"];
            newCreative.DeclaredClickThroughUrls = (IList<string>) parsedArgs[
                "declared_click_urls"];
            newCreative.DeclaredRestrictedCategories = (IList<string>) parsedArgs[
                "declared_restricted_categories"];
            newCreative.DeclaredVendorIds = (IList<int?>) parsedArgs["declared_vendor_ids"];
            newCreative.Video = videoContent;

            BuyersResource.CreativesResource.CreateRequest request =
                rtbService.Buyers.Creatives.Create(newCreative, parent);
            Creative response = null;

            Console.WriteLine("Creating video creative for buyer: {0}", parent);

            try
            {
                response = request.Execute();
            }
            catch (System.Exception exception)
            {
                throw new ApplicationException(
                    $"Real-time Bidding API returned error response:\n{exception.Message}");
            }

            Utilities.PrintCreative(response);
        }
    }
}
